[Inductor][Triton][FP8] Support tile-wise (1x128) scaling in Inductor

jananisriram opened the pull request:
Summary:
Support tile-wise `1x128` scaling in Inductor Triton for FP8 GEMMs, i.e. scaling values along tensors `a` and `b` represent a `1x128` slice of input.

NOTE: Block-wise `128x128` and `1x128` scaling is only supported in CUDA 12.9+; therefore, tile-wise scaling is currently unsupported in `fbcode` (CUDA 12.4). Use OSS PyTorch to run tile-wise scaling (as with deepseek-style scaling).

Test Plan:
Works out-of-the-box with TritonBench:
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 buck2 run mode/{opt,inplace} pytorch/tritonbench:run -- --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 256 --n 768 --k 512 --output="/home/jananisriram/personal/random_bench.csv" --scaling-pair=BlockWise1x128,BlockWise1x128 --atol=1e-2 --rtol=0.5
```

Differential Revision: D84025878




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

drisspg reviewed (approved):
Much cleaner, a couple small comments. Great job!

njriasan reviewed (approved):
@jananisriram See my feedback. Overall this seems good, but I have some suggestions to have a single point of failure for the allowed scaling options. I'll defer to you though regarding what combinations should actually be supported and how to enforce those.

jananisriram commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1253,6 +1475,93 @@ def tuned_sparse_semi_structured_mm(
     )
 
 
+scaling_pairs = [
```
this is the limited set of pairs that are legal for now (since these are the only combinations supported in aten) and seems to be as similar as we can get to the aten approach. by "encoding in a single variable," do you mean that we assign some variable to each pair, like `tensorwise_scaling = (TensorWise, TensorWise)`? i feel that the latter assignment will get complicated, considering that it's potentially possible that we support other combinations in the future

jananisriram commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1253,6 +1475,93 @@ def tuned_sparse_semi_structured_mm(
+    for scale_option_a, scale_option_b in scaling_pairs:
+        if is_desired_scaling(
+            mat_a, scale_a_size, scale_option_a
+        ) and is_desired_scaling(mat_b, scale_b_size, scale_option_b, transpose=True):
```
we need to verify the stride values based on the transposed version of the second tensor, like aten's current implementation; it's possible, however, that this is a function of max-autotune enablement

jananisriram commented on torch/_meta_registrations.py:
```diff
@@ -6444,6 +6444,20 @@ def ceil_div(a, b):
+            elif (
+                scale_a.size(0) == m
+                and scale_a.size(1) == scale_b.size(0) == (_k + 128 - 1) // 128
+                and scale_b.size(1) == (n + 128 - 1) // 128
```
good point, thanks! moved the `ceil_div` function to a location where i can use it here as well

jananisriram committed:
[Inductor][Triton][FP8] Support tile-wise (1x128) scaling in Inductor (#165132)

Summary:

Support tile-wise `1x128` scaling in Inductor Triton for FP8 GEMMs, i.e. scaling values along tensors `a` and `b` represent a `1x128` slice of input.

NOTE: Block-wise `128x128` and `1x128` scaling is only supported in CUDA 12.9+; therefore, tile-wise scaling is currently unsupported in `fbcode` (CUDA 12.4). Use OSS PyTorch to run tile-wise scaling (as with deepseek-style scaling).

Test Plan:
Works out-of-the-box with TritonBench:
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 buck2 run mode/{opt,inplace} pytorch/tritonbench:run -- --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 256 --n 768 --k 512 --output="/home/jananisriram/personal/random_bench.csv" --scaling-pair=BlockWise1x128,BlockWise1x128 --atol=1e-2 --rtol=0.5
```

Reviewed By: njriasan

Differential Revision: D84025878

jeanschmidt commented:
@pytorchbot merge -i
